.gitbook/contents/developer/features/aa-basics/hybrid-compute.md

```diff
@@ -0,0 +1,549 @@
+        dec = ethabi.decode(['uint32', 'uint32'], req['reqBytes'])
+
+        if dec[0] >= dec[1]:
+            s = dec[0] + dec[1]
+            d = dec[0] - dec[1]
+            resp = ethabi.encode(['uint256', 'uint256'], [s, d])
+            err_code = 0
+        else:
+            print("offchain_addsub2 underflow error", dec[0], dec[1])
+            resp = Web3.to_bytes(text="underflow error")
+    except Exception as e:
+        print("DECODE FAILED", e)
+
+    return gen_response(req, err_code, resp)
+```
+
+First things first, we initialize an ``err_code`` and a resp object with values in case of an exception:
+
+``` python
+err_code = 1
+resp = Web3.to_bytes(text="unknown error")
+```
+
+In the try-block we parse the request with the help of this function:
+
+```python
+def parse_req(sk, src_addr, src_nonce, oo_nonce, payload):
+    req = dict()
+    req['skey'] = Web3.to_bytes(hexstr=sk)
+    req['srcAddr'] = Web3.to_checksum_address(src_addr)
+    req['srcNonce'] = Web3.to_int(hexstr=src_nonce)
+    req['opNonce'] = Web3.to_int(hexstr=oo_nonce)
+    req['reqBytes'] = Web3.to_bytes(hexstr=payload)
+    return req
+```
+
+and decode the ``reqBytes`` to an array of ``[uin32, uint32]`` since we want to receive two numbers.
+
+``` python
+dec = ethabi.decode(['uint32', 'uint32'], req['reqBytes'])
+```
+
+Now we can perform our custom logic on the received values:
+
+```python
+if dec[0] >= dec[1]:
+```
+
+When the calculation was successful, we overwrite the previously created err_code and resp variables with successful
+values:
+
+```python
+resp = ethabi.encode(['uint256', 'uint256'], [s, d])
+err_code = 0
+```
+
+As we decoded the ``reqBytes`` by letting the decode function know that we want to receive two numbers by giving it an
+array
+of ``['uin32', 'uint32']``, we let the encode function know, on how we want to decode both numbers. In this case
+with ``['uint256', 'uint256']``.
+
+In case of an underflow error, we encode a string containing a text to let the user know what happened:
+
+```python
+resp = Web3.to_bytes(text="underflow error")
+```
+
+Before we can return these objects, we need to transform them into a specific object:
+
+``` python
+def gen_response(req, err_code, resp_payload):
+    resp2 = ethabi.encode(['address', 'uint256', 'uint32', 'bytes'], [req['srcAddr'], req['srcNonce'], err_code, resp_payload])
+
+    enc1 = ethabi.encode(['bytes32', 'bytes'], [req['skey'], resp2])
+
+    p_enc1 = "0x" + selector("PutResponse(bytes32,bytes)") + \
+        Web3.to_hex(enc1)[2:]  # dfc98ae8
+
+    enc2 = ethabi.encode(['address', 'uint256', 'bytes'], [
+                         Web3.to_checksum_address(HelperAddr), 0, Web3.to_bytes(hexstr=p_enc1)])
+    p_enc2 = "0x" + selector("execute(address,uint256,bytes)") + \
+        Web3.to_hex(enc2)[2:]  # b61d27f6
+
+    limits = {
+        'verificationGasLimit': "0x10000",
+        'preVerificationGas': "0x10000",
+    }
+    callGas = 705*len(resp_payload) + 170000
```

Comment:
Note that this gas calculation is subject to change. I will be adding a version flag to the RPC request so that we can manage changes to the protocol, and moving some of this code into a library so that users won't have to deal with all of the complexity directly.

Reply:
noted, thx!